packages/shared-ini-file-loader/src/slurpFile.ts

```diff
@@ -0,0 +1,62 @@
+// ToDo: Change to "fs/promises" when supporting nodejs>=14
+import { promises as fsPromises } from "fs";
+
+const { readFile, stat } = fsPromises;
+
+type callbacks = { resolve: Function; reject: Function };
+type fileStatus = {
+  contents: string;
+  isReading: boolean;
+  lastModified: Date;
+  requestQueue: callbacks[];
+};
+
+const fileReadHash: { [key: string]: fileStatus } = {};
+
+const callReadFile = (path: string, cbs: callbacks) => {
+  fileReadHash[path].requestQueue.push(cbs);
+  readFile(path, "utf8")
+    .then((data) => {
+      // File read successful
+      fileReadHash[path].contents = data;
+      fileReadHash[path].requestQueue.forEach(({ resolve }) => resolve(data));
+    })
+    .catch((err) => {
+      fileReadHash[path].requestQueue.forEach(({ reject }) => reject(err));
+    })
+    .finally(() => {
+      fileReadHash[path].isReading = false;
+      fileReadHash[path].requestQueue = [];
+      // Update last modified.
+      stat(path).then(({ mtime }) => {
+        fileReadHash[path].lastModified = mtime;
+      });
+    });
+};
+
+export const slurpFile = (path: string) =>
+  new Promise<string>((resolve, reject) => {
+    if (!fileReadHash[path]) {
+      // File not read yet, set file isReading to true and read file.
+      fileReadHash[path] = { isReading: true, lastModified: new Date(0), contents: "", requestQueue: [] };
```

Comment:
`lastModified` may confuces maintainers. The SDK already implemented the memoization for ALL the config loading https://github.com/aws/aws-sdk-js-v3/blob/40e600e598e88afa28f15459b4f9574dcfd171a4/packages/node-config-provider/src/configLoader.ts#L31-L37
Moreover, all the non-credential configs are static value. It means the memoized value withint the client will persist forever, I don't think `slurpFile` will be called again for a memoized config.

Reply:
👍 like your approach

Reply:
Superceded by promise callback queue for loadSharedConfigFiles() in https://github.com/aws/aws-sdk-js-v3/pull/3282